### PR TITLE
feat: Adds GoogleLogger to Diagnostics.Common

### PR DESCRIPTION
**Pull request by amanda-tarafa**
As always, one commit at a time, although I'll rebase everything into one before merging.

We can make more AspNetCore* types obsolete, and I also want to improve docs, but I think this is now good for a beta release.

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
(build error is because of the language version missmatch, I'll figure it out, but this is all green on local so it's ready for review).

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs [hunk 1]
```diff
@@ -545,11 +547,11 @@ public async Task Logging_Trace_External_OneEntry(TestServer server)
                 LogEntry entry = Assert.Single(results);
 
                 // And the resource name of the trace associated to it contains the external trace id.
-                Assert.Contains(TestEnvironment.GetTestProjectId(), entry.Trace);
-                Assert.Contains("external_trace_id", entry.Trace);
+                      Assert.Contains(TestEnvironment.GetTestProjectId(), entry.Trace);
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs:550]: Nit: indentation

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs [hunk 2]
```diff
@@ -441,23 +443,20 @@ public void Log_NoFormatParams()
...
+            ulong spanId = 1;
             string fullTraceName = TraceTarget.ForProject(ProjectId).GetFullTraceName(traceId);
 
             Predicate<IEnumerable<LogEntry>> matcher = logEntries =>
             {
                 LogEntry entry = logEntries.Single();
                 return entry.LogName == new LogName(ProjectId, BaseLogName).ToString() &&
                     entry.Trace == fullTraceName &&
-                    entry.SpanId == spanId;
+                    entry.SpanId == $"{spanId:x16}";
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs:456]: It's unclear to me why this had to change. Is it because we're not requiring the external span IDs to be 64-bit integers?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs [hunk 3]
```diff
@@ -33,6 +33,7 @@ namespace Google.Cloud.Diagnostics.AspNetCore
     /// These values can be attached to a log entry to establish the
     /// relation of it and a trace.
     /// </summary>
+    [Obsolete("Use Common.ITraceContext instead.")]
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs:36]: Maybe fully-qualify this name?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs [hunk 4]
```diff
@@ -67,10 +67,12 @@ private string Formatter(string message, Exception ex)
             Common.LoggerOptions options = Common.LoggerOptions.CreateWithServiceContext(
                 logLevel, logName, labels, monitoredResource, retryOptions: retryOptions, serviceName: serviceName, version: version);
 #pragma warning disable CS0618 // Type or member is obsolete
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs:69]: Do we still need this pragma?

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/KeyValuePairEnumerableExtensions.cs [hunk 5]
```diff
@@ -23,7 +23,7 @@ namespace Google.Cloud.Diagnostics.Common
     /// Extension methods for converting KeyValuePair Enumerables to several types,
     /// including Protobuf well known types.
     /// </summary>
-    public static class KeyValuePairEnumerableExtensions
+    internal static class KeyValuePairEnumerableExtensions
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/KeyValuePairEnumerableExtensions.cs:26]: This isn't a breaking change because it was introduced since the last release, yes?

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/GoogleLogger.cs [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        internal GoogleLogger(
+            IConsumer<LogEntry> consumer, LogTarget logTarget, LoggerOptions loggerOptions,
+            string logName, IClock clock = null, IServiceProvider serviceProvider = null)
+#pragma warning disable CS0618 // Type or member is obsolete
+            : this (consumer, logTarget, loggerOptions, logName, null, null, clock, serviceProvider)
+#pragma warning restore CS0618 // Type or member is obsolete
+        {
+        }
+
+        [Obsolete("Added for backward compatibility only when moving GoogleLogger to Common.")]
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/GoogleLogger.cs:73]: I'm somewhat lost as to whether we actually still need this. Given that it's internal, I'd expect to be able to remove it?

**Review by jskeet — APPROVED** [review]
Okay, I'm glad you've got a handle on this... it *looks* okay to me, but it's slightly hard to follow what's going on, despite your evident care in creating the commits separately... it's just inherently complex.

Looks fine as far as I can see... if you're confident in it, let's merge (once the language issues have been fixed) and release a beta.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs (hunk 4 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs:69]: Yes, this is the one that is used to keep backward compatibility.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerTest.cs:456]: Yes, the now obsolete TraceContextForLogEntry has a string for SpanId, and docs saying that it should be in base 16. But now we use the new ITraceContext where span ID is a ulong, and for setting it in the entry we are the ones who have to convert it to base 16.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs]: Done

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/KeyValuePairEnumerableExtensions.cs (hunk 5 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/KeyValuePairEnumerableExtensions.cs:26]: Yes, that's just one from the previous PR.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/GoogleLogger.cs (hunk 6 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/GoogleLogger.cs:73]: This is the one remaining piece that allows us to use the Common.GoogleLogger from the AspNetCore.GoogleLogger taking into account some of the obsolte types that remained in AspNetCore* that were not moved to Common. This is called from Common.GoogleLoggleProvider which is called from AspNetCore.GoogleLoggerProvider. I added the Obsolete attribute so that we ourselves know not to depend on it.

**Review by amanda-tarafa — COMMENTED** [review]
Addressed all comments and fixed the language version mismatch. Will wait on green for merging.

**Comment by jskeet** [comment]
Looks like there are a couple of doc-comment issues and some compatibility issues. I'm happy to look at the compatibility ones if you'd like.

**Comment by amanda-tarafa** [comment]
I'm looking, but the doc comments one are not a problem on local, so I'm not sure what's happening.

**Comment by amanda-tarafa** [comment]
Doc-comment issues fixed.

The compaibility differences I'm pretty sure are not a problem, but a second pair of eyes would be great. I think is just the tool that's being more strict than it needs to.

**Comment by amanda-tarafa** [comment]
```
Interface 'ILogEntryLabelProvider' implemented interface added 'ILogEntryLabelProvider'
Interface 'ILogEntryLabelProvider'; method 'void Invoke(Dictionary<String, String> labels)' removed.
```
These are:
- `AspNetCore.ILogEntryLabelProvider`  now implementes `Common.ILogEntryLabelProvider`.
- The method was removed from `AspNetCore.ILogEntryLabelProvider` because it's now in `Common.ILogEntryLabelProvider`.

**Comment by jskeet** [comment]
Merging with admin rights. Although the breaking changes *are* theoretically breaking, we don't expect them to actually cause any issues in this case. Will note this in the release notes.
